File: api/mqtt.py
"""
MQTT client for Home Assistant integration.

Publishes GLM state and subscribes to control commands.
Supports Home Assistant MQTT Discovery for automatic entity creation.
"""
import json
import logging
import threading
import time
from typing import Optional, Callable

import paho.mqtt.client as mqtt

from glm_core import SetVolume, AdjustVolume, SetMute, SetDim, SetPower, QueuedAction

logger = logging.getLogger(__name__)

# Default topic prefix
DEFAULT_TOPIC_PREFIX = "glm"

# Home Assistant MQTT Discovery prefix
HA_DISCOVERY_PREFIX = "homeassistant"


class MqttClient:
    """MQTT client for GLM control and state publishing."""

    def __init__(
        self,
        action_queue,
        glm_controller,
        broker: str = "localhost",
        port: int = 1883,
        username: Optional[str] = None,
        password: Optional[str] = None,
        topic_prefix: str = DEFAULT_TOPIC_PREFIX,
        ha_discovery: bool = True,
    ):
        """
        Initialize MQTT client.

        Args:
            action_queue: Queue for submitting GlmActions
            glm_controller: GlmController instance for state
            broker: MQTT broker hostname
            port: MQTT broker port
            username: Optional username for authentication
            password: Optional password for authentication
            topic_prefix: Prefix for all topics (default: "glm")
            ha_discovery: Enable Home Assistant MQTT Discovery
        """
        self._action_queue = action_queue
        self._glm_controller = glm_controller
        self._broker = broker
        self._port = port
        self._username = username
        self._password = password
        self._topic_prefix = topic_prefix
        self._ha_discovery = ha_discovery
        self._client: Optional[mqtt.Client] = None
        self._connected = False
        self._stop_event = threading.Event()

        # Topics
        self._state_topic = f"{topic_prefix}/state"
        self._availability_topic = f"{topic_prefix}/availability"
        self._cmd_volume_topic = f"{topic_prefix}/set/volume"
        self._cmd_mute_topic = f"{topic_prefix}/set/mute"
        self._cmd_dim_topic = f"{topic_prefix}/set/dim"
        self._cmd_power_topic = f"{topic_prefix}/set/power"

    def _on_connect(self, client, userdata, flags, rc, properties=None):
        """Handle connection to broker."""
        if rc == 0:
            logger.info(f"Connected to MQTT broker {self._broker}:{self._port}")
            self._connected = True

            # Subscribe to command topics
            client.subscribe(self._cmd_volume_topic)
            client.subscribe(self._cmd_mute_topic)
            client.subscribe(self._cmd_dim_topic)
            client.subscribe(self._cmd_power_topic)
            logger.info(f"Subscribed to {self._topic_prefix}/set/+")

            # Publish availability
            client.publish(self._availability_topic, "online", retain=True)

            # Publish HA Discovery configs
            if self._ha_discovery:
                self._publish_ha_discovery()

            # Publish current state
            self._publish_state(self._glm_controller.get_state())
        else:
            logger.error(f"Failed to connect to MQTT broker, rc={rc}")

    def _on_disconnect(self, client, userdata, rc, properties=None):
        """Handle disconnection from broker."""
        self._connected = False
        if rc != 0:
            logger.warning(f"Unexpected MQTT disconnect, rc={rc}")

    def _on_message(self, client, userdata, msg):
        """Handle incoming MQTT messages."""
        topic = msg.topic
        try:
            payload = msg.payload.decode('utf-8').strip()
        except UnicodeDecodeError:
            logger.warning(f"Invalid payload on {topic}")
            return

        logger.debug(f"MQTT received: {topic} = {payload}")

        try:
            if topic == self._cmd_volume_topic:
                # Volume: accept integer 0-127
                value = int(payload)
                self._submit_action(SetVolume(target=value))

            elif topic == self._cmd_mute_topic:
                # Mute: accept ON/OFF/true/false/1/0/TOGGLE
                state = self._parse_bool_or_toggle(payload)
                self._submit_action(SetMute(state=state))

            elif topic == self._cmd_dim_topic:
                # Dim: accept ON/OFF/true/false/1/0/TOGGLE
                state = self._parse_bool_or_toggle(payload)
                self._submit_action(SetDim(state=state))

            elif topic == self._cmd_power_topic:
                # Power: only toggle supported
                self._submit_action(SetPower())

        except (ValueError, TypeError) as e:
            logger.warning(f"Invalid MQTT command on {topic}: {payload} ({e})")

    def _parse_bool_or_toggle(self, payload: str) -> Optional[bool]:
        """Parse payload to bool or None (toggle)."""
        payload_lower = payload.lower()
        if payload_lower in ('on', 'true', '1'):
            return True
        elif payload_lower in ('off', 'false', '0'):
            return False
        elif payload_lower == 'toggle':
            return None
        else:
            raise ValueError(f"Unknown state: {payload}")

    def _submit_action(self, action):
        """Submit an action to the queue."""
        self._action_queue.put(QueuedAction(action=action, timestamp=time.time()))

    def _publish_state(self, state: dict):
        """Publish current state to MQTT."""
        if not self._connected or self._client is None:
            return

        # Convert to HA-friendly format
        payload = json.dumps({
            "volume": state["volume"],
            "mute": "ON" if state["mute"] else "OFF",
            "dim": "ON" if state["dim"] else "OFF",
            "power": "ON" if state["power"] else "OFF",
        })

        self._client.publish(self._state_topic, payload, retain=True)

    def _publish_ha_discovery(self):
        """Publish Home Assistant MQTT Discovery configuration."""
        if self._client is None:
            return

        device_info = {
            "identifiers": ["glm_controller"],
            "name": "Genelec GLM",
            "manufacturer": "Genelec",
            "model": "GLM Controller",
        }

        # Volume as number entity
        volume_config = {
            "name": "GLM Volume",
            "unique_id": "glm_volume",
            "command_topic": self._cmd_volume_topic,
            "state_topic": self._state_topic,
            "value_template": "{{ value_json.volume }}",
            "min": 0,
            "max": 127,
            "step": 1,
            "icon": "mdi:volume-high",
            "availability_topic": self._availability_topic,
            "device": device_info,
        }
        self._client.publish(
            f"{HA_DISCOVERY_PREFIX}/number/glm_volume/config",
            json.dumps(volume_config),
            retain=True
        )

        # Mute as switch entity
        mute_config = {
            "name": "GLM Mute",
            "unique_id": "glm_mute",
            "command_topic": self._cmd_mute_topic,
            "state_topic": self._state_topic,
            "value_template": "{{ value_json.mute }}",
            "payload_on": "ON",
            "payload_off": "OFF",
            "icon": "mdi:volume-mute",
            "availability_topic": self._availability_topic,
            "device": device_info,
        }
        self._client.publish(
            f"{HA_DISCOVERY_PREFIX}/switch/glm_mute/config",
            json.dumps(mute_config),
            retain=True
        )

        # Dim as switch entity
        dim_config = {
            "name": "GLM Dim",
            "unique_id": "glm_dim",
            "command_topic": self._cmd_dim_topic,
            "state_topic": self._state_topic,
            "value_template": "{{ value_json.dim }}",
            "payload_on": "ON",
            "payload_off": "OFF",
            "icon": "mdi:brightness-6",
            "availability_topic": self._availability_topic,
            "device": device_info,
        }
        self._client.publish(
            f"{HA_DISCOVERY_PREFIX}/switch/glm_dim/config",
            json.dumps(dim_config),
            retain=True
        )

        # Power as switch entity
        power_config = {
            "name": "GLM Power",
            "unique_id": "glm_power",
            "command_topic": self._cmd_power_topic,
            "state_topic": self._state_topic,
            "value_template": "{{ value_json.power }}",
            "payload_on": "ON",
            "payload_off": "OFF",
            "icon": "mdi:power",
            "availability_topic": self._availability_topic,
            "device": device_info,
        }
        self._client.publish(
            f"{HA_DISCOVERY_PREFIX}/switch/glm_power/config",
            json.dumps(power_config),
            retain=True
        )

        logger.info("Published Home Assistant MQTT Discovery configs")

    def on_state_change(self, state: dict):
        """Callback for GlmController state changes."""
        self._publish_state(state)

    def start(self):
        """Start the MQTT client."""
        self._client = mqtt.Client(mqtt.CallbackAPIVersion.VERSION2)

        if self._username:
            self._client.username_pw_set(self._username, self._password)

        # Set last will for availability
        self._client.will_set(self._availability_topic, "offline", retain=True)

        # Set callbacks
        self._client.on_connect = self._on_connect
        self._client.on_disconnect = self._on_disconnect
        self._client.on_message = self._on_message

        # Connect
        try:
            self._client.connect(self._broker, self._port, keepalive=60)
        except Exception as e:
            logger.error(f"Failed to connect to MQTT broker: {e}")
            return

        # Start network loop in background thread
        self._client.loop_start()
        logger.info(f"MQTT client started, connecting to {self._broker}:{self._port}")

        # Register state callback
        self._glm_controller.add_state_callback(self.on_state_change)

    def stop(self):
        """Stop the MQTT client."""
        if self._client:
            # Unregister callback
            self._glm_controller.remove_state_callback(self.on_state_change)

            # Publish offline status
            if self._connected:
                self._client.publish(self._availability_topic, "offline", retain=True)

            self._client.loop_stop()
            self._client.disconnect()
            logger.info("MQTT client stopped")


def start_mqtt_client(
    action_queue,
    glm_controller,
    broker: str,
    port: int = 1883,
    username: Optional[str] = None,
    password: Optional[str] = None,
    topic_prefix: str = DEFAULT_TOPIC_PREFIX,
    ha_discovery: bool = True,
) -> MqttClient:
    """
    Start MQTT client for Home Assistant integration.

    Args:
        action_queue: Queue for submitting GlmActions
        glm_controller: GlmController instance
        broker: MQTT broker hostname
        port: MQTT broker port
        username: Optional username
        password: Optional password
        topic_prefix: Topic prefix (default: "glm")
        ha_discovery: Enable HA MQTT Discovery

    Returns:
        MqttClient instance
    """
    client = MqttClient(
        action_queue=action_queue,
        glm_controller=glm_controller,
        broker=broker,
        port=port,
        username=username,
        password=password,
        topic_prefix=topic_prefix,
        ha_discovery=ha_discovery,
    )
    client.start()
    return client

File: go/mqtt/mqtt.go
// Package mqtt provides an MQTT client for Home Assistant integration.
//
// Publishes GLM state and subscribes to control commands.
// Supports Home Assistant MQTT Discovery for automatic entity creation.
package mqtt

import (
	"encoding/json"
	"fmt"
	"log/slog"
	"strconv"
	"strings"
	"time"

	pahomqtt "github.com/eclipse/paho.mqtt.golang"

	"vol20toglm/controller"
	"vol20toglm/types"
)

const haDiscoveryPrefix = "homeassistant"

// Client is an MQTT client for GLM control and state publishing.
type Client struct {
	client    pahomqtt.Client
	ctrl      *controller.Controller
	actions   chan<- types.Action
	traceGen  *types.TraceIDGenerator
	log       *slog.Logger
	topicPfx  string
	haDiscov  bool
}

// statePayload is the JSON published to the state topic.
type statePayload struct {
	Volume   int    `json:"volume"`
	VolumeDB int    `json:"volume_db"`
	Mute     string `json:"mute"`
	Dim      string `json:"dim"`
	Power    string `json:"power"`
}

func boolToOnOff(b bool) string {
	if b {
		return "ON"
	}
	return "OFF"
}

// Start creates and connects the MQTT client. Returns nil if broker is empty.
func Start(
	broker string,
	port int,
	username, password string,
	topicPrefix string,
	haDiscovery bool,
	ctrl *controller.Controller,
	actions chan<- types.Action,
	traceGen *types.TraceIDGenerator,
	log *slog.Logger,
) *Client {
	if broker == "" {
		return nil
	}

	c := &Client{
		ctrl:     ctrl,
		actions:  actions,
		traceGen: traceGen,
		log:      log,
		topicPfx: topicPrefix,
		haDiscov: haDiscovery,
	}

	opts := pahomqtt.NewClientOptions()
	opts.AddBroker(fmt.Sprintf("tcp://%s:%d", broker, port))
	opts.SetClientID("vol20toglm")
	opts.SetAutoReconnect(true)
	opts.SetWill(topicPrefix+"/availability", "offline", 0, true)
	opts.SetOnConnectHandler(c.onConnect)
	opts.SetConnectionLostHandler(c.onConnectionLost)

	if username != "" {
		opts.SetUsername(username)
		opts.SetPassword(password)
	}

	c.client = pahomqtt.NewClient(opts)

	token := c.client.Connect()
	token.Wait()
	if err := token.Error(); err != nil {
		log.Error("MQTT connect failed", "broker", broker, "port", port, "err", err)
		return nil
	}

	log.Info("MQTT client started", "broker", broker, "port", port, "topic", topicPrefix)

	// Register state callback so changes are published automatically
	ctrl.OnStateChange(func(_, new_ types.State) {
		c.publishState(new_)
	})

	return c
}

// Stop disconnects the client gracefully.
func (c *Client) Stop() {
	if c == nil || c.client == nil {
		return
	}
	c.client.Publish(c.topicPfx+"/availability", 0, true, "offline")
	c.client.Disconnect(1000)
	c.log.Info("MQTT client stopped")
}

func (c *Client) onConnect(client pahomqtt.Client) {
	c.log.Info("MQTT connected")

	// Subscribe to command topics
	cmdTopics := map[string]pahomqtt.MessageHandler{
		c.topicPfx + "/set/volume": c.handleVolume,
		c.topicPfx + "/set/mute":   c.handleMute,
		c.topicPfx + "/set/dim":    c.handleDim,
		c.topicPfx + "/set/power":  c.handlePower,
	}
	for topic, handler := range cmdTopics {
		client.Subscribe(topic, 0, handler)
	}
	c.log.Info("MQTT subscribed", "topics", c.topicPfx+"/set/+")

	// Publish availability
	client.Publish(c.topicPfx+"/availability", 0, true, "online")

	// Publish HA Discovery configs
	if c.haDiscov {
		c.publishHADiscovery()
	}

	// Publish current state
	c.publishState(c.ctrl.GetState())
}

func (c *Client) onConnectionLost(_ pahomqtt.Client, err error) {
	c.log.Warn("MQTT connection lost", "err", err)
}

// Command handlers

func (c *Client) handleVolume(_ pahomqtt.Client, msg pahomqtt.Message) {
	payload := strings.TrimSpace(string(msg.Payload()))
	tid := c.traceGen.Next("mqtt")
	c.log.Debug("MQTT rx", "topic", msg.Topic(), "payload", payload, "trace_id", tid)

	value, err := strconv.Atoi(payload)
	if err != nil {
		c.log.Warn("MQTT invalid volume", "payload", payload, "trace_id", tid, "err", err)
		return
	}
	// Accept dB value (-127 to 0) or raw value (0-127)
	if value <= 0 {
		value = value + 127
	}
	if value < 0 {
		value = 0
	} else if value > 127 {
		value = 127
	}

	c.actions <- types.Action{
		Kind:      types.KindSetVolume,
		Value:     value,
		Source:    "mqtt",
		TraceID:   tid,
		Timestamp: time.Now(),
	}
}

func (c *Client) handleMute(_ pahomqtt.Client, msg pahomqtt.Message) {
	payload := strings.TrimSpace(string(msg.Payload()))
	tid := c.traceGen.Next("mqtt")
	c.log.Debug("MQTT rx", "topic", msg.Topic(), "payload", payload, "trace_id", tid)

	boolVal, toggle, err := parseBoolOrToggle(payload)
	if err != nil {
		c.log.Warn("MQTT invalid mute", "payload", payload, "trace_id", tid, "err", err)
		return
	}

	c.actions <- types.Action{
		Kind:      types.KindSetMute,
		BoolValue: boolVal,
		Toggle:    toggle,
		Source:    "mqtt",
		TraceID:   tid,
		Timestamp: time.Now(),
	}
}

func (c *Client) handleDim(_ pahomqtt.Client, msg pahomqtt.Message) {
	payload := strings.TrimSpace(string(msg.Payload()))
	tid := c.traceGen.Next("mqtt")
	c.log.Debug("MQTT rx", "topic", msg.Topic(), "payload", payload, "trace_id", tid)

	boolVal, toggle, err := parseBoolOrToggle(payload)
	if err != nil {
		c.log.Warn("MQTT invalid dim", "payload", payload, "trace_id", tid, "err", err)
		return
	}

	c.actions <- types.Action{
		Kind:      types.KindSetDim,
		BoolValue: boolVal,
		Toggle:    toggle,
		Source:    "mqtt",
		TraceID:   tid,
		Timestamp: time.Now(),
	}
}

func (c *Client) handlePower(_ pahomqtt.Client, msg pahomqtt.Message) {
	payload := strings.TrimSpace(string(msg.Payload()))
	tid := c.traceGen.Next("mqtt")
	c.log.Debug("MQTT rx", "topic", msg.Topic(), "payload", payload, "trace_id", tid)

	boolVal, toggle, err := parseBoolOrToggle(payload)
	if err != nil {
		c.log.Warn("MQTT invalid power", "payload", payload, "trace_id", tid, "err", err)
		return
	}

	c.actions <- types.Action{
		Kind:      types.KindSetPower,
		BoolValue: boolVal,
		Toggle:    toggle,
		Source:    "mqtt",
		TraceID:   tid,
		Timestamp: time.Now(),
	}
}

// State publishing

func (c *Client) publishState(state types.State) {
	if c.client == nil || !c.client.IsConnected() {
		return
	}

	payload := statePayload{
		Volume:   state.Volume,
		VolumeDB: state.Volume - 127,
		Mute:     boolToOnOff(state.Mute),
		Dim:      boolToOnOff(state.Dim),
		Power:    boolToOnOff(state.Power),
	}

	data, err := json.Marshal(payload)
	if err != nil {
		c.log.Error("MQTT marshal state failed", "err", err)
		return
	}

	c.client.Publish(c.topicPfx+"/state", 0, true, data)
}

// Home Assistant MQTT Discovery

func (c *Client) publishHADiscovery() {
	device := map[string]interface{}{
		"identifiers":  []string{"glm_controller"},
		"name":         "Genelec GLM",
		"manufacturer": "Genelec",
		"model":        "GLM Controller",
	}

	avail := c.topicPfx + "/availability"
	state := c.topicPfx + "/state"

	// Volume as number entity (dB: -127 to 0)
	c.publishDiscoveryConfig("number", "glm_volume", map[string]interface{}{
		"name":                "GLM Volume",
		"unique_id":           "glm_volume",
		"command_topic":       c.topicPfx + "/set/volume",
		"state_topic":         state,
		"value_template":      "{{ value_json.volume_db }}",
		"min":                 -127,
		"max":                 0,
		"step":                1,
		"unit_of_measurement": "dB",
		"icon":                "mdi:volume-high",
		"availability_topic":  avail,
		"device":              device,
	})

	// Mute as switch (config category — keeps device toggle = Power only)
	c.publishDiscoveryConfig("switch", "glm_mute", map[string]interface{}{
		"name":               "GLM Mute",
		"unique_id":          "glm_mute",
		"command_topic":      c.topicPfx + "/set/mute",
		"state_topic":        state,
		"value_template":     "{{ value_json.mute }}",
		"payload_on":         "ON",
		"payload_off":        "OFF",
		"icon":               "mdi:volume-mute",
		"availability_topic": avail,
		"entity_category":    "config",
		"device":             device,
	})

	// Dim as switch (config category — keeps device toggle = Power only)
	c.publishDiscoveryConfig("switch", "glm_dim", map[string]interface{}{
		"name":               "GLM Dim",
		"unique_id":          "glm_dim",
		"command_topic":      c.topicPfx + "/set/dim",
		"state_topic":        state,
		"value_template":     "{{ value_json.dim }}",
		"payload_on":         "ON",
		"payload_off":        "OFF",
		"icon":               "mdi:brightness-6",
		"availability_topic": avail,
		"entity_category":    "config",
		"device":             device,
	})

	// Power as switch
	c.publishDiscoveryConfig("switch", "glm_power", map[string]interface{}{
		"name":               "GLM Power",
		"unique_id":          "glm_power",
		"command_topic":      c.topicPfx + "/set/power",
		"state_topic":        state,
		"value_template":     "{{ value_json.power }}",
		"payload_on":         "ON",
		"payload_off":        "OFF",
		"icon":               "mdi:power",
		"availability_topic": avail,
		"device":             device,
	})

	c.log.Info("MQTT HA Discovery configs published")
}

func (c *Client) publishDiscoveryConfig(component, objectID string, config map[string]interface{}) {
	topic := fmt.Sprintf("%s/%s/%s/config", haDiscoveryPrefix, component, objectID)
	data, err := json.Marshal(config)
	if err != nil {
		c.log.Error("MQTT marshal discovery config failed", "component", component, "err", err)
		return
	}
	c.client.Publish(topic, 0, true, data)
}

// Helpers

func parseBoolOrToggle(payload string) (boolVal bool, toggle bool, err error) {
	switch strings.ToLower(payload) {
	case "on", "true", "1":
		return true, false, nil
	case "off", "false", "0":
		return false, false, nil
	case "toggle":
		return false, true, nil
	default:
		return false, false, fmt.Errorf("unknown state: %s", payload)
	}
}
